{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# LMO 2013 - 2014 analysis\n",
    "\n",
    "Not much was given to me on the dataset, except the metadata file:\n",
    "\n",
    "> I think now Johannes has removed the reads from the spike-in control\n",
    "(T. thermophilus) from the Baltic transect samples, so you could\n",
    "probably run CRASS on those now Sergiu (if Johannes sends a link).\n",
    "On the other hand, it would be interesting to include the spike-in and\n",
    "see if CRASS can predict the crisprs present in the genome. According to\n",
    "the crispr database T. thermophilus DSM7039 (I think this is the strain\n",
    "they used as spike-in) should have 2 crispr clusters on the chromosome\n",
    "and 8 on a plasmid:\n",
    "\n",
    "http://crispr.i2bc.paris-saclay.fr/cgi-bin/crispr/SpecieProperties_db.cgi?Taxon_id[]=262724\n",
    "\n",
    "(strain DSM7039 should be the same as HB27:\n",
    "https://www.dsmz.de/catalogues/details/culture/DSM-7039.html) \n",
    "\n",
    "\n",
    "I used these samples here:\n",
    "/proj/b2014214/nobackup/preprocessing/fastuniq/quality_adapter_no_indices/P4201_1*_R{1,2}.fq.gz\n",
    "\n",
    "Johannes processed the P4201 files but not the ones I have (I asked for the unprocessed) \n",
    "I used the cutadapt arguments:\n",
    "-n 3 --minimum-length 31 -a AGATCGGAAGAGCACACGTCTGAACTCCAGTCAC -G ^CGTGTGCTCTTCCGATCT -A AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGT\n",
    "That removed around 5-10% of the bases.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "import sys\n",
    "import pandas as pd\n",
    "import matplotlib\n",
    "matplotlib.style.use('ggplot')\n",
    "import seaborn as sns\n",
    "sns.set(color_codes=True)\n",
    "\n",
    "# laptop config:\n",
    "# source code location\n",
    "#src = \"/home/sergiu/data/work/andersson/src/andersson/src\"\n",
    "# crispr file generated by crass (xml format)\n",
    "#crispr_location = \"/home/sergiu/data/work/andersson/data/asko/crass/crass.crispr\"\n",
    "# tab separated output\n",
    "#crispr_tab_location = \"/home/sergiu/data/work/andersson/data/asko/spacer_table_filtered_all.txt\"\n",
    "\n",
    "#desktop config:\n",
    "# source code location\n",
    "src = \"/home/sergiu/data/data/work/andersson/src/andersson/src\"\n",
    "# crispr file generated by crass (xml format)\n",
    "crispr_location = \"/home/sergiu/data/data/work/andersson/data/lmo2013-2014/crass/crass.crispr\"\n",
    "# tab separated output\n",
    "crispr_tab_location = \"/home/sergiu/data/data/work/andersson/data/lmo2013-2014/lmo_spacer_table_filtered_all.txt\"\n",
    "\n",
    "\n",
    "sample_list = [\n",
    "'P4201_112_R1', 'P4201_122_R1', 'P4201_101_R1', 'P4201_106_R1', \n",
    "'P4201_120_R1', 'P4201_118_R1', 'P4201_111_R1', 'P4201_110_R1', \n",
    "'P4201_121_R1', 'P4201_107_R1', 'P4201_119_R1', 'P4201_104_R1', \n",
    "'P4201_124_R1', 'P4201_105_R1', 'P4201_109_R1', 'P4201_108_R1', \n",
    "'P4201_114_R1', 'P4201_123_R1', 'P4201_116_R1', 'P4201_117_R1', \n",
    "'P4201_102_R1', 'P4201_103_R1',\n",
    "'P4201_112_R2', 'P4201_122_R2', 'P4201_101_R2', 'P4201_106_R2', \n",
    "'P4201_120_R2', 'P4201_118_R2', 'P4201_111_R2', 'P4201_110_R2', \n",
    "'P4201_121_R2', 'P4201_107_R2', 'P4201_119_R2', 'P4201_104_R2', \n",
    "'P4201_124_R2', 'P4201_105_R2', 'P4201_109_R2', 'P4201_108_R2', \n",
    "'P4201_114_R2', 'P4201_123_R2', 'P4201_116_R2', 'P4201_117_R2', \n",
    "'P4201_102_R2', 'P4201_103_R2']\n",
    "\n",
    "# copy-pasted from SRA\n",
    "meta = {'SRR3727523': '29-Jun-2011', 'SRR3727518': '30-Aug-2011', \n",
    "        'SRR3727515': '14-Jun-2011', 'SRR3727513': '17-Aug-2011',\n",
    "        'SRR3727509': '03-Aug-2011', 'SRR3727505': '20-Jul-2011'}\n",
    "\n",
    "timeorder = sample_list\n",
    "\n",
    "if not src in sys.path:\n",
    "    sys.path.append(src)\n",
    "else:\n",
    "    print(sys.path)\n",
    "\n",
    "import crass_crispr_parser as crispr"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Generate the CRISPR table\n",
    "\n",
    "This step was sourced."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:andersson]",
   "language": "python",
   "name": "conda-env-andersson-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
